src/caustic/cosmology.py

```diff
@@ -34,127 +36,102 @@
-        return self.comoving_dist(z) / (1 + z)
-
-    def angular_diameter_dist_z1z2(self, z1, z2):
-        """
-        [Mpc]
-        """
-        return self.comoving_dist_z1z2(z1, z2) / (1 + z2)
-
-    def time_delay_dist(self, z_l, z_s):
-        """
-        [Mpc]
-        """
-        d_l = self.angular_diameter_dist(z_l)
-        d_s = self.angular_diameter_dist(z_s)
-        d_ls = self.angular_diameter_dist_z1z2(z_l, z_s)
+    def comoving_dist_z1z2(
+        self, z1: Tensor, z2: Tensor, x: dict[str, Any] = defaultdict(list)) -> Tensor:
+        return self.comoving_dist(z2, x) - self.comoving_dist(z1, x)
+
+    def angular_diameter_dist(
+        self, z: Tensor, x: dict[str, Any] = defaultdict(list)) -> Tensor:
+        return self.comoving_dist(z, x) / (1 + z)
+
+    def angular_diameter_dist_z1z2(
+        self, z1: Tensor, z2: Tensor, x: dict[str, Any] = defaultdict(list)) -> Tensor:
+        return self.comoving_dist_z1z2(z1, z2, x) / (1 + z2)
+
+    def time_delay_dist(
+        self, z_l: Tensor, z_s: Tensor, x: dict[str, Any] = defaultdict(list)) -> Tensor:
+        d_l = self.angular_diameter_dist(z_l, x)
+        d_s = self.angular_diameter_dist(z_s, x)
+        d_ls = self.angular_diameter_dist_z1z2(z_l, z_s, x)
         return (1 + z_l) * d_l * d_s / d_ls
 
-    def Sigma_cr(self, z_l, z_s):
-        """
-        Critical lensing density [solMass / Mpc^2]
-        """
-        d_l = self.angular_diameter_dist(z_l)
-        d_s = self.angular_diameter_dist(z_s)
-        d_ls = self.angular_diameter_dist_z1z2(z_l, z_s)
+    def Sigma_cr(
+        self, z_l: Tensor, z_s: Tensor, x: dict[str, Any] = defaultdict(list)) -> Tensor:
+        d_l = self.angular_diameter_dist(z_l, x)
+        d_s = self.angular_diameter_dist(z_s, x)
+        d_ls = self.angular_diameter_dist_z1z2(z_l, z_s, x)
         return d_s / d_l / d_ls / (4 * pi * G_over_c2)
 
 
-class FlatLambdaCDMCosmology(Cosmology):
+class FlatLambdaCDM(Cosmology):
     """
-    Flat LCDM cosmology with no radiation.
 
-    Notes:
-        Compute comoving_dist on the fly
+    Flat LCDM cosmology with no radiation.
     """
 
     def __init__(
         self,
-        h0=h0_default,
-        rho_cr_0=rho_cr_0_default,
-        Om0=Om0_default,
-        device: torch.device = torch.device("cpu"),
-        dtype: torch.dtype = torch.float32,
+        name: str,
+        h0: Optional[Tensor] = torch.tensor(h0_default),
+        rho_cr_0: Optional[Tensor] = torch.tensor(rho_cr_0_default),
+        Om0: Optional[Tensor] = torch.tensor(Om0_default),
     ):
-        super().__init__(h0, device, dtype)
-        self.rho_cr_0 = rho_cr_0
-        self.Om0 = Om0
-        self.Ode0 = 1 - Om0
+        super().__init__(name)
+
+        self.add_param("h0", h0)
+        self.add_param("rho_cr_0", rho_cr_0)
+        self.add_param("Om0", Om0)
 
         self._comoving_dist_helper_x_grid = _comoving_dist_helper_x_grid.to(
-            self.device, dtype
+            dtype=torch.float32
```

Comment:
I'm envisioning that we'll put `Parametrized` objects on GPU the same way as `nn.Module`s: first you create the `Parametrized`, then you call `.to("cuda")`. Otherwise we'll need to add a `device` arg to `__init__` and call `.to(device)` in there, which seems redundant to me.